File: PISC/potentials/Quartic_bistable.py
import numpy as np
from PISC.potentials.base import PES
from numpy import exp
from PISC.potentials.quartic_bistable_f import quartic_bistable as quartic_bistable_f

class quartic_bistable(PES):
        def __init__(self,alpha,D,lamda,g,z,invert_pes=1.0):
            super(quartic_bistable).__init__()
            """
            This is the two-dimensional double well potential with a 1D double well potential along the x-axis 
             a 1D Morse potential along the y-axis coupled with a nonlinear term 

             D: Dissociation energy of the Morse potential
             alpha: Morse potential parameter (determines the width of the potential well, appears in the exponent)
             
             g: Double well parameter g as in Hashimoto's paper
             lamda: Double well parameter lambda as in Hashimoto's paper

             z: Coupling strength
            """
            
            self.D = D
            self.alpha = alpha

            self.g = g
            self.lamda = lamda
            
            self.z = z  
    
            self.param_list = [alpha,D,lamda,g,z]

        def bind(self,ens,rp,pes_fort=False,transf_fort=False):
            super(quartic_bistable,self).bind(ens,rp,pes_fort,transf_fort=transf_fort)
 
        def potential(self,q):
            x = q[:,0] 
            y = q[:,1]
            return self.potential_xy(x,y) 
        
        def dpotential(self,q):
            x = q[:,0] 
            y = q[:,1]
            Vx = 4*self.g*x*(-1 + np.exp(-self.alpha*y*self.z))*(x**2 - self.lamda**2/(8*self.g)) + 4*self.g*x*(x**2 - self.lamda**2/(8*self.g))
            Vy = 2*self.D*self.alpha*(1 - np.exp(-self.alpha*y))*np.exp(-self.alpha*y) \
                    - self.alpha*self.z*(self.g*(x**2 - self.lamda**2/(8*self.g))**2 - self.lamda**4/(64*self.g))*np.exp(-self.alpha*y*self.z)       
            
            return self.dpot_rearrange([Vx,Vy])

        def ddpotential(self,q):
            x = q[:,0] 
            y = q[:,1]
            Vxx = 4*self.g*(-2*x**2*(1 - np.exp(-self.alpha*y*self.z)) + 3*x**2 - (1 - np.exp(-self.alpha*y*self.z))*(8*x**2 - self.lamda**2/self.g)/8 - self.lamda**2/(8*self.g))
            Vxy = -4*self.alpha*self.g*x*self.z*(x**2 - self.lamda**2/(8*self.g))*np.exp(-self.alpha*y*self.z)
            Vyy = self.alpha**2*(-2*self.D*(1 - np.exp(-self.alpha*y))*np.exp(-self.alpha*y) + 2*self.D*np.exp(-2*self.alpha*y) \
                        + self.z**2*(self.g*(8*x**2 - self.lamda**2/self.g)**2 - self.lamda**4/self.g)*np.exp(-self.alpha*y*self.z)/64)

            return self.ddpot_rearrange([[Vxx,Vxy],[Vxy,Vyy]])

        def potential_xy(self,x,y):
            eay = np.exp(-self.alpha*y)
            quartx = (x**2 - self.lamda**2/(8*self.g))
            vy = self.D*(1-eay)**2
            vx = self.g*quartx**2
            vxy = (vx-self.lamda**4/(64*self.g))*(np.exp(-self.z*self.alpha*y) - 1) 
            return vx + vy + vxy

        def potential_f(self,qcart_f,pot_f):
            return quartic_bistable_f.potential_f(qcart_f,self.param_list,pot_f) 
            
        def dpotential_f(self,qcart_f,dpot_f):
            return quartic_bistable_f.dpotential_f(qcart_f,self.param_list,dpot_f)

        def ddpotential_f(self,qcart_f,ddpot_f):
            return quartic_bistable_f.ddpotential_f(qcart_f,self.param_list,ddpot_f)

File: PISC/potentials/Quartic_bistable.f90
module quartic_bistable
        implicit none
        private
        public :: potential, dpotential, ddpotential
        public :: potential_f, dpotential_f, ddpotential_f
        contains
        subroutine potential(q,ndim,param_list,lp,Vq)
                integer, intent(in) :: ndim, lp
                real(8), intent(in) :: q(ndim)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(out) :: Vq
                real(8) :: x,y, alpha, D, lamda, g, z
                real(8) :: ea, eay, quartx, vy, vx, vxy
                        
                alpha = param_list(1)
                D = param_list(2)
                lamda = param_list(3)
                g = param_list(4)
                z = param_list(5)

                x = q(1)
                y = q(2)
                
                eay = exp(-alpha*y)
                quartx = (x**2 - lamda**2/(8*g))
                vy = D*(1-eay)**2
                vx = g*quartx**2
                vxy = (vx-lamda**4/(64*g))*(exp(-z*alpha*y) - 1)

                Vq = vx + vy + vxy

        end subroutine potential

        subroutine dpotential(q,ndim,param_list,lp,dVq)
                integer, intent(in) :: ndim, lp
                real(8), intent(in) :: q(ndim)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: dVq(ndim)
                real(8) :: x,y, alpha, D, lamda, g, z
                real(8) :: ea, eay, quartx, vy, vx, vxy
                        
                alpha = param_list(1)
                D = param_list(2)
                lamda = param_list(3)
                g = param_list(4)
                z = param_list(5)

                x = q(1)
                y = q(2)
                
                dVq(1) = 4*g*x*(-1 + exp(-alpha*y*z))*(x**2 - lamda**2/(8*g)) + 4*g*x*(x**2 - lamda**2/(8*g)) 
                dVq(2) = 2*D*alpha*(1 - exp(-alpha*y))*exp(-alpha*y) - &
                        alpha*z*(g*(x**2 - lamda**2/(8*g))**2 - lamda**4/(64*g))*exp(-alpha*y*z) 

        end subroutine dpotential

        subroutine ddpotential(q,ndim,param_list,lp,ddVq)
                integer, intent(in) :: ndim, lp
                real(8), intent(in) :: q(ndim)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: ddVq(ndim,ndim)
                real(8) :: x,y, alpha, D, lamda, g, z
                real(8) :: ea, eay, quartx, vy, vx, vxy
                        
                alpha = param_list(1)
                D = param_list(2)
                lamda = param_list(3)
                g = param_list(4)
                z = param_list(5)
                
                x = q(1)
                y = q(2)
                
                ddVq(1,1) = 4*g*(-2*x**2*(1 - exp(-alpha*y*z)) + 3*x**2 - &
                            (1 - exp(-alpha*y*z))*(8*x**2 - lamda**2/g)/8 - lamda**2/(8*g))
                ddVq(1,2) = -4*alpha*g*x*z*(x**2 - lamda**2/(8*g))*exp(-alpha*y*z) 
                ddVq(2,2) = alpha**2*(-2*D*(1 - exp(-alpha*y))*exp(-alpha*y) + 2*D*exp(-2*alpha*y) &
                            + z**2*(g*(8*x**2 - lamda**2/g)**2 - lamda**4/g)*exp(-alpha*y*z)/64) 
                ddVq(2,1) = ddVq(1,2)

        end subroutine ddpotential
 
        subroutine potential_f(qcart,N,ndim,nbeads,param_list,lp,pot)
                integer, intent(in) :: N, ndim, nbeads, lp
                real(8), intent(in) :: qcart(nbeads,ndim,N)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: pot(nbeads,N)
                integer :: i,j

                !!$omp parallel do private(i,j) shared(qcart,pot)
                do i = 1, N
                        do j = 1, nbeads
                                call potential(qcart(j,:,i),ndim,param_list,lp,pot(j,i))
                        end do
                end do
                !!$omp end parallel do

        end subroutine potential_f

        subroutine dpotential_f(qcart,N,ndim,nbeads,param_list,lp,dpot_cart)
                integer, intent(in) :: N, ndim, nbeads, lp
                real(8), intent(in) :: qcart(nbeads,ndim,N)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: dpot_cart(nbeads,ndim,N)
                integer :: i,j

                !!$omp parallel do private(i,j) shared(qcart,dpot_cart)
                do i = 1, N
                        do j = 1, nbeads
                                call dpotential(qcart(j,:,i),ndim,param_list,lp,dpot_cart(j,:,i))
                        end do
                end do
                !!$omp end parallel do

        end subroutine dpotential_f
                
        subroutine ddpotential_f(qcart,N,ndim,nbeads,param_list,lp,ddpot_cart)
                integer, intent(in) :: N, ndim, nbeads, lp
                real(8), intent(in) :: qcart(nbeads,ndim,N)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: ddpot_cart(nbeads,ndim,nbeads,ndim,N)
                integer :: i,j

                !!$omp parallel do private(i,j) shared(qcart,ddpot_cart)
                do i=1, N
                        do j=1, nbeads
                                call ddpotential(qcart(j,:,i),ndim,param_list,lp,ddpot_cart(j,:,j,:,i))
                        end do
                end do
                !!$omp end parallel do

        end subroutine ddpotential_f
end module quartic_bistable
